manticore/ethereum.py

```diff
@@ -655,17 +661,49 @@ def make_symbolic_value(self):
         return ABI.SValue
 
     @staticmethod
-    def compile(source_code, contract_name=None):
+    def compile(source_code, contract_name=None, libraries=None):
         ''' Get initialization bytecode from a Solidity source code '''
-        name, source_code, bytecode, runtime, srcmap, srcmap_runtime, hashes, abi, warnings = ManticoreEVM._compile(source_code, contract_name)
+        name, source_code, bytecode, runtime, srcmap, srcmap_runtime, hashes, abi, warnings = ManticoreEVM._compile(source_code, contract_name, libraries)
         return bytecode
 
     @staticmethod
-    def _compile(source_code, contract_name):
+    def _link(bytecode, libraries=None):
+        has_dependencies = '_' in bytecode
+        hex_contract = bytecode
+        if has_dependencies:
+            deps = {}
+            pos = 0
+            while pos < len(hex_contract):
+                if hex_contract[pos] == '_':
+                    # __/tmp/tmp_9k7_l:Manticore______________
+                    lib_placeholder = hex_contract[pos:pos + 40]
+                    lib_name = lib_placeholder.split(':')[1].split('_')[0]
+                    deps.setdefault(lib_name, []).append(pos)
+                    pos += 40
+                else:
+                    pos += 2
+
+            if libraries is None:
+                raise DependencyError(deps.keys())
+            libraries = dict(libraries)
+            hex_contract_lst = list(hex_contract)
```

Comment:
We are changing it to list just because a string doesn't support item assignment, is that right?

Maybe we should use bytearray instead for that purpose? An example:
```
>>> a = bytearray('x'*10, 'utf8')
>>> a[2:4] = b'xy'
>>> a[5:6] = [1, 2]
>>> a[-1] = 123
>>> a
bytearray(b'xxxyx\x01\x02xxx{')
>>> bytes(a)
b'xxxyx\x01\x02xxx{'
```